ambry-api/src/main/java/com/github/ambry/compression/Compression.java

```diff
@@ -226,4 +228,86 @@ default byte[] decompress(byte[] compressedBuffer, int compressedBufferOffset, i
         originalData, 0, originalData.length);
     return originalData;
   }
+
+  /**
+   * Get the original data size stored inside the shared/composite compressed buffer.  The compressed buffer
+   * contains version, algorithm name, original data size, and compressed data.  This method helps to determine
+   * the decompressed buffer size before calling decompress().
+   * This method does not change the index of compressedBuffer.
+   *
+   * @param compressedBuffer The compressed buffer.  The buffer must be position to the beginning of buffer.
+   *                         The indexes are not affected and remain unchanged.
+   * @return The size of original data.
+   * @throws CompressionException Throws this exception if compression has internal failure.
+   */
+  int getDecompressBufferSize(ByteBuffer compressedBuffer) throws CompressionException;
+
+  /**
+   * Compress the buffer specified in {@code sourceData}.  The entire sourceBuffer will be read,
+   * so set the position and limit to the range of data to read in sourceBuffer before calling this method.
+   * After calling this method, sourceBuffer position will be advanced to the buffer limit,
+   * and the compressedBuffer position will be advanced to the end of the compressed binary.
+   *
+   * @param sourceBuffer The source/uncompressed data to compress.  It cannot be null or empty.
+   * @param compressedBuffer The compressed buffer where the compressed data will be written to.  Its size must be
+   *                       at least the size return from getCompressBufferSize() or it may fail due to buffer size.
+   * @return The actual compressed data size in bytes.
+   * @throws CompressionException Throws this exception if compression has internal failure.
+   */
+  int compress(ByteBuffer sourceBuffer, ByteBuffer compressedBuffer) throws CompressionException;
+
+  /**
+   * Compress the buffer specified in {@code sourceBuffer}.  This method allocates either direct memory or heap memory,
+   * based on the allocateDirectBuffer parameter, to store the compressed data.  The output compressed buffer capacity
+   * may be bigger than necessary in case sourceBuffer is incompressible.  The index of sourceBuffer will be updated
```

Comment:
why would the compressed buffer be bigger than the sourceBuffer if its not compressed ?

Reply:
This is algorithm specific thing.   For example, if compressing 1-byte, the compressed output is likely be bigger than 1 byte due to structural overhead and internal compression dictionary, depending algorithm.